File: code75/LevelOrder.java
class Node {
    int data;
    Node left;
    Node right;
    
    public Node(int data) {
        this.data = data;
    }
}

class LevelOrder {
    public static void main(String[] args) {
        Node root = new Node(100);
        root.left = new Node(30);
        root.right = new Node(10);
        
        root.left.left = new Node(1);
        root.left.right = new Node(3);
        
        root.right.left = new Node(66);
        root.right.right = new Node(77);

        root.right.left.right = new Node(12345);
        
        int height = calHeight(root);
        printLevelOrder(root, height);
    }

    static int calHeight(Node root) {
        if(root == null) {
            return 0;
        }
        else {
            int l = calHeight(root.left);
            int r = calHeight(root.right);
            if(l>r) {
                return l+1;
            }
            else {
                return r+1;
            }
        }
    }

    static void printLevelOrder(Node root, int h)
    {
        for(int i = 0; i<=h; i++) {
            printlevel(root, i);
        }
    }

    static void printlevel(Node root, int h)
    {
        if(root == null){
            return;
        }
        if(h==1) {
            System.out.print(root.data + " ");
        }
        else if(h>1) {
            printlevel(root.left, h-1);
            printlevel(root.right, h-1);
        }
    }

}

File: Golang/2020Go/tree/levelorder.go
package main

import "fmt"

type Node struct {
	val   int
	left  *Node
	right *Node
}

func heightOfTree(root *Node) int {
	if root == nil {
		return 0
	}

	var left = heightOfTree(root.left)
	var right = heightOfTree(root.right)

	if left > right {
		return left + 1
	} else {
		return right + 1
	}
}

func printLevel(root *Node, h int) {
	if root == nil {
		return
	}

	if h == 1 {
		fmt.Printf("%d ", root.val)
		return
	}
	if h > 1 {
		printLevel(root.left, h-1)

		printLevel(root.right, h-1)
	}
}

func levelOrder(root *Node) {
	if root == nil {
		return
	}
	var height = heightOfTree(root)
	fmt.Printf("Height of a tree : %d \n", height)
	for i := 1; i <= height; i++ {
		printLevel(root, i)
	}
}

func main() {
	// BST Representation
	//         100
	//        /   \
	//      50    150
	//     / \    /  \
	//   20  80  110 200
	//  /  \
	// 10  30

	root := Node{100, nil, nil}
	root.left = &Node{50, nil, nil}
	root.right = &Node{150, nil, nil}
	root.left.left = &Node{20, nil, nil}
	root.left.left.left = &Node{10, nil, nil}
	root.left.left.right = &Node{30, nil, nil}
	root.left.right = &Node{80, nil, nil}
	root.right.left = &Node{110, nil, nil}
	root.right.right = &Node{200, nil, nil}

	fmt.Println("Level Order Traversal : ")
	levelOrder(&root)

}
